text/0314-cdk-construct-hub.md

```diff
@@ -0,0 +1,278 @@
+header of all other pages), a list of relevant results is displayed as tiles,
+similar to the short-list of packages shown on the landing page.
+
+### Package Detail Page
+
+> [Example rendering](#rendering-package-detail-page)
+
+Clicking on a given package's tile in the landing page or search result page
+brings the user to a package detail page that represents the following
+information:
+
+- Package name, tags, and description
+- Overview of publishing information (author, version, license, etc...)
+  - Where possible, links to the source repository, issue tracker, etc...
+- Install instructions
+- Available languages
+- Package's documentation
+  - Rendered content of the package's `README.md`
+  - API reference documentation, in all supported languages
+
+The list of available languages is also a selector: upon clicking a language,
+the installation instructions and documentation automatically switch to the
+relevant form for the currently selected language.
+
+### Support
+
+The CDK Construct Hub offers ways for customers to engage maintainers about
+various problems, including the following:
+
+- A malicious package was identified
+- Incorrectly rendered documentation
+- Package un-listing request
+
+When a contact request originates from a package detail page, the customer is
+guided to the package's issue tracker or public repository if their issue is
+with the package itself (and not about it's listing in CDK Construct Hub).
+
+## Technical Design
+
+### What is the high level implementation plan?
+
+#### Overview
+
+The CDK Construct Hub is a relatively simple static web application. While the
+dataset it exposes (all CDK Construct Libraries) may be relatively fast moving,
+there is no necessity for newly published packages (or versions thereof) to be
+indexed and presented on the application particularly quickly. A consistency
+window between a couple of hours and a day is probably acceptable.
+
+A simple data pipeline will ingest package publication events created by a
+source-specific adapter (e.g: by polling on the CouchDB instance replica for the
+[npmjs.com registry](https://skimdb.npmjs.com/registry)), and produce static
+artifacts the website is based off of. This pipeline is in particular
+responsible for extracting the `.jsii` assembly file from the npm packages, and
+to prefetch all information necessary to back the webiste's search and package
+detail pages.
+
+#### Back-End
+
+1. The back-end for CDK Construct Hub starts with a custom-built event source,
+   which implements the logic necessary to detect new packages that are relevant
+   to the CDK Construct Hub, and sends messages to an SQS queue for further
+   processing. Notifications have the following attributes:
+
+   Name        | Description
+   ------------|----------------------------------------------------------------
+   `assembly`  | The `.jsii` assembly included in the package
```

Comment:
The assembly can be quite big. Will this be an issue?

Reply:
I guess it may be necessary to stage it into S3 in certain extreme cases... Or compression? Both?

Reply:
Since we need the .jsii file in S3 anyway (this is the primary source of the UI), I think a reference to it should be fine.

Reply:
I like the reference idea but I don't really think it will matter much.  Even Oracle can do JSON well now.

Reply:
It's mostly an implementation detail, I think.

Reply:
I think it's safe to assume that some .jsii manifests cannot fit into an SQS message (how big is the aws-cdk-lib manifest?).

Would it make more sense for this stage to just enqueue a message that contains the URL of the tarball and have the rest of the pipeline handle downloading the tarball and extracting the .jsii manifest from it?